lib/ast_ml.ml

```diff
@@ -0,0 +1,134 @@
+module type S_P_TREE = sig
```

Comment:
What are the tradeoffs behind splitting each data type into its own module? Why not just provide the datatypes natively in the S_P_TREE module? 

For that matter, what is the goal of the S_P_TREE module type?